bug: The mysqld is crashed when you are starting replication.

### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/stoneatom/stonedb/blob/stonedb-5.7-dev/CONTRIBUTING.md).

### Please confirm if bug report does NOT exists already ?

- [X] I confirm there is no existing issue for this

### Describe the problem

```sql
Program received signal SIGABRT, Aborted.
[Switching to Thread 0x7f0337fff700 (LWP 36871)]
0x00007f133e593387 in raise () from /lib64/libc.so.6
(gdb) bt
#0  0x00007f133e593387 in raise () from /lib64/libc.so.6
#1  0x00007f133e594a78 in abort () from /lib64/libc.so.6
#2  0x00007f133e58c1a6 in __assert_fail_base () from /lib64/libc.so.6
#3  0x00007f133e58c252 in __assert_fail () from /lib64/libc.so.6
#4  0x0000000002e4c446 in Tianmu::mm::TraceableObject::~TraceableObject (this=0x7f0310000a30, __in_chrg=<optimized out>) at /home/Code/GitHub/stonedb/storage/tianmu/mm/traceable_object.cpp:131
#5  0x0000000002d9dae2 in Tianmu::core::Pack::~Pack (this=0x7f0310000a30, __in_chrg=<optimized out>) at /home/Code/GitHub/stonedb/storage/tianmu/core/pack.h:45
#6  0x0000000003096524 in Tianmu::core::PackInt::PackInt (this=0x7f0310000a30, dpn=0x7ef042201e98, pc=..., s=0x7ef272668bc0) at /home/Code/GitHub/stonedb/storage/tianmu/core/pack_int.cpp:33
#7  0x0000000002db4c18 in __gnu_cxx::new_allocator<Tianmu::core::PackInt>::construct<Tianmu::core::PackInt<Tianmu::core::DPN*&, Tianmu::core::ObjectId<(Tianmu::core::COORD_TYPE)0, 3, Tianmu::core::object_id_helper::empty> const&, Tianmu::core::ColumnShare*&> > (this=0x7f0337ffd83f, __p=0x7f0310000a30, __args#0=@0x7f0337ffda28: 0x7ef042201e98, __args#1=..., __args#2=@0x7f058c02b618: 0x7ef272668bc0) at /opt/rh/devtoolset-8/root/usr/include/c++/8/ext/new_allocator.h:136
#8  0x0000000002db371e in std::allocator_traits<std::allocator<Tianmu::core::PackInt> >::construct<Tianmu::core::PackInt<Tianmu::core::DPN*&, Tianmu::core::ObjectId<(Tianmu::core::COORD_TYPE)0, 3, Tianmu::core::object_id_helper::empty> const&, Tianmu::core::ColumnShare*&> > (__a=..., __p=0x7f0310000a30, __args#0=@0x7f0337ffda28: 0x7ef042201e98, __args#1=..., __args#2=@0x7f058c02b618: 0x7ef272668bc0)
    at /opt/rh/devtoolset-8/root/usr/include/c++/8/bits/alloc_traits.h:475
#9  0x0000000002db0eb2 in std::_Sp_counted_ptr_inplace<Tianmu::core::PackInt, std::allocator<Tianmu::core::PackInt>, (__gnu_cxx::_Lock_policy)2>::_Sp_counted_ptr_inplace<Tianmu::core::DPN*&, Tianmu::core::ObjectId<(Tianmu::core::COORD_TYPE)0, 3, Tianmu::core::object_id_helper::empty> const&, Tianmu::core::ColumnShare*&> (this=0x7f0310000a20, __a=...) at /opt/rh/devtoolset-8/root/usr/include/c++/8/bits/shared_ptr_base.h:545
#10 0x0000000002dad91d in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<Tianmu::core::PackInt, std::allocator<Tianmu::core::PackInt>, Tianmu::core::DPN*&, Tianmu::core::ObjectId<(Tianmu::core::COORD_TYPE)0, 3, Tianmu::core::object_id_helper::empty> const&, Tianmu::core::ColumnShare*&> (this=0x7f0337ffda48, __p=@0x7f0337ffda40: 0x0, __a=...) at /opt/rh/devtoolset-8/root/usr/include/c++/8/bits/shared_ptr_base.h:677
#11 0x0000000002da9c64 in std::__shared_ptr<Tianmu::core::PackInt, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<std::allocator<Tianmu::core::PackInt>, Tianmu::core::DPN*&, Tianmu::core::ObjectId<(Tianmu::core::COORD_TYPE)0, 3, Tianmu::core::object_id_helper::empty> const&, Tianmu::core::ColumnShare*&> (this=0x7f0337ffda40, __tag=...) at /opt/rh/devtoolset-8/root/usr/include/c++/8/bits/shared_ptr_base.h:1342
#12 0x0000000002da70d9 in std::shared_ptr<Tianmu::core::PackInt>::shared_ptr<std::allocator<Tianmu::core::PackInt>, Tianmu::core::DPN*&, Tianmu::core::ObjectId<(Tianmu::core::COORD_TYPE)0, 3, Tianmu::core::object_id_helper::empty> const&, Tianmu::core::ColumnShare*&> (this=0x7f0337ffda40, __tag=...) at /opt/rh/devtoolset-8/root/usr/include/c++/8/bits/shared_ptr.h:359
#13 0x0000000002da4b5d in std::allocate_shared<Tianmu::core::PackInt, std::allocator<Tianmu::core::PackInt>, Tianmu::core::DPN*&, Tianmu::core::ObjectId<(Tianmu::core::COORD_TYPE)0, 3, Tianmu::core::object_id_helper::empty> const&, Tianmu::core::ColumnShare*&> (__a=..., __args#0=@0x7f0337ffda28: 0x7ef042201e98, __args#1=..., __args#2=@0x7f058c02b618: 0x7ef272668bc0) at /opt/rh/devtoolset-8/root/usr/include/c++/8/bits/shared_ptr.h:706
#14 0x0000000002da19ff in std::make_shared<Tianmu::core::PackInt, Tianmu::core::DPN*&, Tianmu::core::ObjectId<(Tianmu::core::COORD_TYPE)0, 3, Tianmu::core::object_id_helper::empty> const&, Tianmu::core::ColumnShare*&> (
    __args#0=@0x7f0337ffda28: 0x7ef042201e98, __args#1=..., __args#2=@0x7f058c02b618: 0x7ef272668bc0) at /opt/rh/devtoolset-8/root/usr/include/c++/8/bits/shared_ptr.h:722
#15 0x0000000002d94193 in Tianmu::core::TianmuAttr::Fetch (this=0x7f058c02b480, pc=...) at /home/Code/GitHub/stonedb/storage/tianmu/core/tianmu_attr.cpp:870
#16 0x0000000002da1489 in Tianmu::core::DataCache::GetOrFetchObject<Tianmu::core::Pack, Tianmu::core::ObjectId<(Tianmu::core::COORD_TYPE)0, 3, Tianmu::core::object_id_helper::empty>, Tianmu::core::TianmuAttr> (this=0xa1a05fa0, 
    coord_=..., fetcher_=0x7f058c02b480) at /home/Code/GitHub/stonedb/storage/tianmu/core/data_cache.h:234
#17 0x0000000002d938b6 in Tianmu::core::TianmuAttr::LockPackForUse (this=0x7f058c02b480, pn=88) at /home/Code/GitHub/stonedb/storage/tianmu/core/tianmu_attr.cpp:797
#18 0x0000000002d968ba in Tianmu::core::TianmuAttr::<lambda()>::operator()(void) const (__closure=0x7f0337ffded0) at /home/Code/GitHub/stonedb/storage/tianmu/core/tianmu_attr.cpp:1200
#19 0x0000000002d9adf7 in std::_Function_handler<void(), Tianmu::core::TianmuAttr::DeleteBatchData(Tianmu::core::Transaction*, const std::vector<long unsigned int>&)::<lambda()> >::_M_invoke(const std::_Any_data &) (__functor=...)
    at /opt/rh/devtoolset-8/root/usr/include/c++/8/bits/std_function.h:297
#20 0x0000000002cc4ed2 in std::function<void ()>::operator()() const (this=0x7f0337ffded0) at /opt/rh/devtoolset-8/root/usr/include/c++/8/bits/std_function.h:687
#21 0x0000000002cfceb7 in Tianmu::core::FunctionExecutor::FunctionExecutor(std::function<void ()>, std::function<void ()>) (this=0x7f0337ffdd50, call_in_constructor=..., call_in_deconstructor=...)
    at /home/Code/GitHub/stonedb/storage/tianmu/core/tools.h:271
#22 0x0000000002d96be1 in Tianmu::core::TianmuAttr::DeleteBatchData (this=0x7f058c02b480, tx=0x7f058c0143f0, rows=std::vector of length 3, capacity 3 = {...}) at /home/Code/GitHub/stonedb/storage/tianmu/core/tianmu_attr.cpp:1200
#23 0x0000000002e02667 in std::__invoke_impl<void, void (Tianmu::core::TianmuAttr::*&)(Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> > const&), Tianmu::core::TianmuAttr<Tianmu::core::Transaction*&, std::vector<unsigned long, std::allocator<unsigned long> >&>*&> (__f=
    @0x7f058c02f2f8: (void (Tianmu::core::TianmuAttr::*)(Tianmu::core::TianmuAttr * const, Tianmu::core::Transaction *, const std::vector<unsigned long, std::allocator<unsigned long> > &)) 0x2d968e6 <Tianmu::core::TianmuAttr::DeleteBatchData(Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> > const&)>, __t=@0x7f058c02f328: 0x7f058c02b480, __args#0=@0x7f058c02f320: 0x7f058c0143f0, __args#1=std::vector of length 3, capacity 3 = {...})
    at /opt/rh/devtoolset-8/root/usr/include/c++/8/bits/invoke.h:73
#24 0x0000000002e01fed in std::__invoke<void (Tianmu::core::TianmuAttr::*&)(Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> > const&), Tianmu::core::TianmuAttr*&, Tianmu::core::Transaction*&, std::vector<unsigned long, std::allocator<unsigned long> >&> (__fn=
    @0x7f058c02f2f8: (void (Tianmu::core::TianmuAttr::*)(Tianmu::core::TianmuAttr * const, Tianmu::core::Transaction *, const std::vector<unsigned long, std::allocator<unsigned long> > &)) 0x2d968e6 <Tianmu::core::TianmuAttr::DeleteBatchData(Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> > const&)>, __args#0=@0x7f058c02f328: 0x7f058c02b480, __args#1=@0x7f058c02f320: 0x7f058c0143f0, 
    __args#2=std::vector of length 3, capacity 3 = {...}) at /opt/rh/devtoolset-8/root/usr/include/c++/8/bits/invoke.h:95
#25 0x0000000002e01768 in std::_Bind<void (Tianmu::core::TianmuAttr::*(Tianmu::core::TianmuAttr*, Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> >))(Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> > const&)>::__call<void, , 0ul, 1ul, 2ul>(std::tuple<>&&, std::_Index_tuple<0ul, 1ul, 2ul>) (this=0x7f058c02f2f8, __args=empty std::tuple)
    at /opt/rh/devtoolset-8/root/usr/include/c++/8/functional:400
#26 0x0000000002dfffd8 in std::_Bind<void (Tianmu::core::TianmuAttr::*(Tianmu::core::TianmuAttr*, Tianmu::core::Transaction*, std::vector<unsigned long, std::allocator<unsigned long> >))(Tianmu::core::Transaction*, std::vector<unsigned ---Type <return> to continue, or q <return> to quit---
```

### Expected behavior

_No response_

### How To Reproduce

_No response_

### Environment

./mysqld  Ver 5.7.36-StoneDB-v1.0.3 for Linux on x86_64 (build-)
build information as follow: 
        Repository address: https://github.com/stoneatom/stonedb.git:stonedb-5.7-dev
        Branch name: stonedb-5.7-dev
        Last commit ID: 4bd09ac
        Last commit time: Date:   Fri Mar 31 09:43:11 2023 +0800
        Build time: Date: Fri Mar 31 14:02:13 CST 2023

### Are you interested in submitting a PR to solve the problem?

- [X] Yes, I will!

```
[2023-04-11 02:14:31.349637] [42238] [WARN] [exception.cpp:42] MSG: Exception: assert failed on v.HasValue() at pack_str.cpp:198, msg: [./zz/tb_port_run_data.tianmu/columns/11/DATA locationInPack: 0].

STACK TRACE BEGIN
        /home/Code/GitHub/stonedb/build/install//bin/mysqld(Tianmu::common::AssertException::AssertException(char const*, char const*, int, std::string const&)+0x1b7) [0x3005171]
        /home/Code/GitHub/stonedb/build/install//bin/mysqld(Tianmu::core::PackStr::UpdateValue(unsigned long, Tianmu::core::Value const&)+0x171) [0x30ccae1]
        /home/Code/GitHub/stonedb/build/install//bin/mysqld(Tianmu::core::TianmuAttr::UpdateBatchData(Tianmu::core::Transaction*, std::unordered_map<unsigned long, std::shared_ptr<Tianmu::core::Value>, std::hash<unsigned long>, std::equal_to<unsigned long>, std::allocator<std::pair<unsigned long const, std::shared_ptr<Tianmu::core::Value> > > > const&)+0x864) [0x2d9d7b6]
        /home/Code/GitHub/stonedb/build/install//bin/mysqld(void std::__invoke_impl<void, void (Tianmu::core::TianmuAttr::*&)(Tianmu::core::Transaction*, std::unordered_map<unsigned long, std::shared_ptr<Tianmu::core::Value>, std::hash<unsigned long>, std::equal_to<unsigned long>, std::allocator<std::pair<unsigned long const, std::shared_ptr<Tianmu::core::Value> > > > const&), Tianmu::core::TianmuAttr*&, Tianmu::core::Transaction*&, std::unordered_map<unsigned long, std::shared_ptr<Tianmu::core::Value>, std::hash<unsigned long>, std::equal_to<unsigned long>, std::allocator<std::pair<unsigned long const, std::shared_ptr<Tianmu::core::Value> > > >&>(std::__invoke_memfun_deref, void (Tianmu::core::TianmuAttr::*&)(Tianmu::core::Transaction*, std::unordered_map<unsigned long, std::shared_ptr<Tianmu::core::Value>, std::hash<unsigned long>, std::equal_to<unsigned long>, std::allocator<std::pair<unsigned long const, std::shared_ptr<Tianmu::core::Value> > > > const&), Tianmu::core::TianmuAttr*&, Tianmu::core::Transaction*&, std::unordered_map<unsigned long, std::shared_ptr<Tianmu::core::Value>, std::hash<unsigned long>, std::equal_to<unsigned long>, std::allocator<std::pair<unsigned long const, std::shared_ptr<Tianmu::core::Value> > > >&)+0x98) [0x2e0ac53]
        /home/Code/GitHub/stonedb/build/install//bin/mysqld(std::__invoke_result<void (Tianmu::core::TianmuAttr::*&)(Tianmu::core::Transaction*, std::unordered_map<unsigned long, std::shared_ptr<Tianmu::core::Value>, std::hash<unsigned long>, std::equal_to<unsigned long>, std::allocator<std::pair<unsigned long const, std::shared_ptr<Tianmu::core::Value> > > > const&), Tianmu::core::TianmuAttr*&, Tianmu::core::Transaction*&, std::unordered_map<unsigned long, std::shared_ptr<Tianmu::core::Value>, std::hash<unsigned long>, std::equal_to<unsigned long>, std::allocator<std::pair<unsigned long const, std::shared_ptr<Tianmu::core::Value> > > >&>::type std::__invoke<void (Tianmu::core::TianmuAttr::*&)(Tianmu::core::Transaction*, std::unordered_map<unsigned long, std::shared_ptr<Tianmu::core::Value>, std::hash<unsigned long>, std::equal_to<unsigned long>, std::allocator<std::pair<unsigned long const, std::shared_ptr<Tianmu::core::Value> > > > const&), Tianmu::core::TianmuAttr*&, Tianmu::core::Transaction*&, std::unordered_map<unsigned long, std::shared_ptr<Tianmu::core::Value>, std::hash<unsigned long>, std::equal_to<unsigned long>, std::allocator<std::pair<unsigned long const, std::shared_ptr<Tianmu::core::Value> > > >&>(void (Tianmu::core::TianmuAttr::*&)(Tianmu::core::Transaction*, std::unordered_map<unsigned long, std::shared_ptr<Tianmu::core::Value>, std::hash<unsigned long>, std::equal_to<unsigned long>, std::allocator<std::pair<unsigned long const, std::shared_ptr<Tianmu::core::Value> > > > const&), Tianmu::core::TianmuAttr*&, Tianmu::core::Transaction*&, std::unordered_map<unsigned long, std::shared_ptr<Tianmu::core::Value>, std::hash<unsigned long>, std::equal_to<unsigned long>, std::allocator<std::pair<unsigned long const, std::shared_ptr<Tianmu::core::Value> > > >&)+0x6f) [0x2e0a5e7]
        /home/Code/GitHub/stonedb/build/install//bin/mysqld(void std::_Bind<void (Tianmu::core::TianmuAttr::*(Tianmu::core::TianmuAttr*, Tianmu::cor
[2023-04-11 02:14:31.349887] [42193] [ERROR] [engine.cpp:947] MSG: Roll back query 'LOAD DELAYED INSERT DATA INTO TABLE zz.tb_port_run_data'
[2023-04-11 02:14:31.349934] [42193] [INFO] [transaction.cpp:90] MSG: rollback transaction 6434514500002719.
[2023-04-11 02:14:31.349946] [42193] [INFO] [tianmu_table.cpp:679] MSG: roll back table zz../zz/tb_port_run_data.tianmu
[2023-04-11 02:14:31.350042] [42193] [ERROR] [ha_my_tianmu.cpp:121] MSG: RunLoader Error: Tianmu internal error:assert failed on v.HasValue() at pack_str.cpp:198, msg: [./zz/tb_port_run_data.tianmu/columns/11/DATA locationInPack: 0]

```

--- a/mysql-test/suite/tianmu/r/issue1523.result
+++ b/mysql-test/suite/tianmu/r/issue1523.result
@@ -1,0 +1,156 @@
+include/master-slave.inc
+[connection master]
+# on master:
+DROP DATABASE IF EXISTS issue1523_test;
+CREATE DATABASE issue1523_test;
+USE issue1523_test;
+include/sync_slave_sql_with_master.inc
+# on slave:
+USE issue1523_test;
+# on master:
+CREATE TABLE t1 (
+`id` int(10) NOT NULL AUTO_INCREMENT,
+`dev_code` varchar(20) NOT NULL,
+`port` smallint(2) DEFAULT '0',
+`server_ip` varchar(20) DEFAULT NULL,
+`work_status` int(1) DEFAULT '2',
+`port_link_car` int(1) NOT NULL DEFAULT '0',
+`last_order_sn` varchar(42) DEFAULT NULL,
+`last_charge_type` smallint(1) DEFAULT NULL,
+`last_charge_mode` int(2) DEFAULT '0',
+`last_status` smallint(1) DEFAULT NULL,
+`last_soc` int(3) DEFAULT NULL,
+`last_car_no` varchar(42) DEFAULT NULL,
+`last_vin` varchar(45) DEFAULT NULL,
+`create_time` timestamp NULL DEFAULT CURRENT_TIMESTAMP,
+`update_time` timestamp NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+PRIMARY KEY (`id`),
+UNIQUE KEY `portindex` (`dev_code`,`port`),
+KEY `dev_code` (`dev_code`)
+) ENGINE=INNODB;
+insert into t1(id,dev_code,port,server_ip,work_status,port_link_car,last_charge_mode,create_time,update_time) values
+(1,'3301081000000019',0,'172.16.171.219',2,0,0,'2018-08-20 13:46:42','2023-04-02 22:10:30');
+insert into t1(id,dev_code,port,server_ip,work_status,port_link_car,last_charge_mode,create_time,update_time) values
+(2,'3301081000000020',0,'172.16.171.219',3,1,1,'2018-08-20 13:46:42','2023-04-02 22:10:30');
+insert into t1(id,dev_code,port,server_ip,work_status,port_link_car,last_charge_mode,create_time,update_time) values
+(3,'3301081000000021',0,'172.16.171.219',4,2,2,'2018-08-20 13:46:42','2023-04-02 22:10:30');
+insert into t1 values
+(4,'3301081000000022',1,'172.16.171.219',2,1,'2023040314582504221842260',5,0,5,100,'JJY3786','LKLA6G1M4HA727506','2018-08-20 13:46:42','2023-04-02 22:10:30');
+insert into t1 values
+(5,'3301081000000023',1,'172.16.171.219',2,1,'2023040314582504221842261',6,0,6,101,'JJY3787','LKLA6G1M4HA727506','2018-08-20 13:46:42','2023-04-02 22:10:30');
+insert into t1 values
+(6,'3301081000000024',1,'172.16.171.219',2,1,'2023040314582504221842262',7,0,7,102,'JJY3788','LKLA6G1M4HA727506','2018-08-20 13:46:42','2023-04-02 22:10:30');
+update t1 set last_car_no='JJY3786' where id=1;
+update t1 set last_car_no=NULL where id=1;
+insert into t1(id,dev_code,port,server_ip,work_status,port_link_car,last_charge_mode,create_time,update_time) values
+(7,'3301081000000025',0,'172.16.171.219',2,0,0,'2018-08-20 13:46:42','2023-04-02 22:10:30');
+select last_car_no from t1 where id=1;
+last_car_no
+NULL
+select count(*) from t1;
+count(*)
+7
+include/sync_slave_sql_with_master.inc
+# on slave:
+select last_car_no from t1 where id=1;
+last_car_no
+NULL
+select count(*) from t1;
+count(*)
+7
+# on master:
+update t1 set last_car_no='JJY3786' where id=1;
+update t1 set last_car_no=NULL where id=1;
+update t1 set last_car_no='JJY3786' where id=1;
+update t1 set last_car_no=NULL where id=1;
+update t1 set last_car_no='JJY3786' where id=1;
+insert into t1(id,dev_code,port,server_ip,work_status,port_link_car,last_charge_mode,create_time,update_time) values
+(8,'3301081000000026',0,'172.16.171.219',2,0,0,'2018-08-20 13:46:42','2023-04-02 22:10:30');
+select last_car_no from t1 where id=1;
+last_car_no
+JJY3786
+select count(*) from t1;
+count(*)
+8
+include/sync_slave_sql_with_master.inc
+# on slave:
+select last_car_no from t1 where id=1;
+last_car_no
+JJY3786
+select count(*) from t1;
+count(*)
+8
+# on master:
+insert into t1(id,dev_code,port,server_ip,work_status,port_link_car,last_charge_mode,create_time,update_time) values
+(9,'3301081000000027',0,'172.16.171.219',2,0,0,'2018-08-20 13:46:42','2023-04-02 22:10:30');
+delete from t1 where id=9;
+insert into t1(id,dev_code,port,server_ip,work_status,port_link_car,last_charge_mode,create_time,update_time) values
+(9,'3301081000000027',0,'172.16.171.219',2,0,0,'2018-08-20 13:46:42','2023-04-02 22:10:30');
+select last_car_no from t1 where id=9;
+last_car_no
+NULL
+select count(*) from t1;
+count(*)
+9
+include/sync_slave_sql_with_master.inc
+# on slave:
+select last_car_no from t1 where id=9;
+last_car_no
+NULL
+select count(*) from t1;
+count(*)
+9
+# on master:
+update t1 set last_car_no=NULL where id=1;
+delete from t1 where id =1;
+select last_car_no from t1 where id=1;
+last_car_no
+select count(*) from t1;
+count(*)
+8
+include/sync_slave_sql_with_master.inc
+# on slave:
+select last_car_no from t1 where id=1;
+last_car_no
+select count(*) from t1;
+count(*)
+8
+# on master:
+insert into t1(id,dev_code,port,server_ip,work_status,port_link_car,last_charge_mode,create_time,update_time) values
+(10,'3301081000000028',0,'172.16.171.219',2,0,0,'2018-08-20 13:46:42','2023-04-02 22:10:30');
+update t1 set last_car_no='JJY3786' where id=10;
+select last_car_no from t1 where id=10;
+last_car_no
+JJY3786
+select count(*) from t1;
+count(*)
+9
+include/sync_slave_sql_with_master.inc
+# on slave:
+select last_car_no from t1 where id=10;
+last_car_no
+JJY3786
+select count(*) from t1;
+count(*)
+9
+# on master:
+insert into t1(id,dev_code,port,server_ip,work_status,port_link_car,last_charge_mode,create_time,update_time) values
+(11,'3301081000000029',0,'172.16.171.219',2,0,0,'2018-08-20 13:46:42','2023-04-02 22:10:30');
+update t1 set last_car_no='JJY3786' where id=11;
+delete from t1 where id =11;
+select last_car_no from t1 where id=11;
+last_car_no
+select count(*) from t1;
+count(*)
+9
+include/sync_slave_sql_with_master.inc
+# on slave:
+select last_car_no from t1 where id=11;
+last_car_no
+select count(*) from t1;
+count(*)
+9
+# on master:
+DROP DATABASE issue1523_test;
+include/sync_slave_sql_with_master.inc
+stop slave;

--- a/mysql-test/suite/tianmu/t/issue1523-slave.opt
+++ b/mysql-test/suite/tianmu/t/issue1523-slave.opt
@@ -1,0 +1,3 @@
+--tianmu_mandatory=ON
+--tianmu_no_key_error=ON
+--tianmu_insert_delayed=1

--- a/mysql-test/suite/tianmu/t/issue1523.test
+++ b/mysql-test/suite/tianmu/t/issue1523.test
@@ -1,0 +1,154 @@
+--source include/have_tianmu.inc
+--source include/have_binlog_format_row.inc
+--disable_warnings
+--source include/master-slave.inc
+--enable_warnings
+
+connection master;
+--echo # on master:
+--disable_warnings
+DROP DATABASE IF EXISTS issue1523_test;
+--enable_warnings
+CREATE DATABASE issue1523_test;
+USE issue1523_test;
+--source include/sync_slave_sql_with_master.inc
+--echo # on slave:
+USE issue1523_test;
+
+connection master;
+--echo # on master:
+CREATE TABLE t1 (
+  `id` int(10) NOT NULL AUTO_INCREMENT,
+  `dev_code` varchar(20) NOT NULL,
+  `port` smallint(2) DEFAULT '0',
+  `server_ip` varchar(20) DEFAULT NULL,
+  `work_status` int(1) DEFAULT '2',
+  `port_link_car` int(1) NOT NULL DEFAULT '0',
+  `last_order_sn` varchar(42) DEFAULT NULL,
+  `last_charge_type` smallint(1) DEFAULT NULL,
+  `last_charge_mode` int(2) DEFAULT '0',
+  `last_status` smallint(1) DEFAULT NULL,
+  `last_soc` int(3) DEFAULT NULL,
+  `last_car_no` varchar(42) DEFAULT NULL,
+  `last_vin` varchar(45) DEFAULT NULL,
+  `create_time` timestamp NULL DEFAULT CURRENT_TIMESTAMP,
+  `update_time` timestamp NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+  PRIMARY KEY (`id`),
+  UNIQUE KEY `portindex` (`dev_code`,`port`),
+  KEY `dev_code` (`dev_code`)
+) ENGINE=INNODB;
+
+insert into t1(id,dev_code,port,server_ip,work_status,port_link_car,last_charge_mode,create_time,update_time) values
+(1,'3301081000000019',0,'172.16.171.219',2,0,0,'2018-08-20 13:46:42','2023-04-02 22:10:30');
+
+insert into t1(id,dev_code,port,server_ip,work_status,port_link_car,last_charge_mode,create_time,update_time) values
+(2,'3301081000000020',0,'172.16.171.219',3,1,1,'2018-08-20 13:46:42','2023-04-02 22:10:30');
+
+insert into t1(id,dev_code,port,server_ip,work_status,port_link_car,last_charge_mode,create_time,update_time) values
+(3,'3301081000000021',0,'172.16.171.219',4,2,2,'2018-08-20 13:46:42','2023-04-02 22:10:30');
+
+insert into t1 values
+(4,'3301081000000022',1,'172.16.171.219',2,1,'2023040314582504221842260',5,0,5,100,'JJY3786','LKLA6G1M4HA727506','2018-08-20 13:46:42','2023-04-02 22:10:30');
+
+insert into t1 values
+(5,'3301081000000023',1,'172.16.171.219',2,1,'2023040314582504221842261',6,0,6,101,'JJY3787','LKLA6G1M4HA727506','2018-08-20 13:46:42','2023-04-02 22:10:30');
+
+insert into t1 values
+(6,'3301081000000024',1,'172.16.171.219',2,1,'2023040314582504221842262',7,0,7,102,'JJY3788','LKLA6G1M4HA727506','2018-08-20 13:46:42','2023-04-02 22:10:30');
+
+update t1 set last_car_no='JJY3786' where id=1;
+update t1 set last_car_no=NULL where id=1;
+insert into t1(id,dev_code,port,server_ip,work_status,port_link_car,last_charge_mode,create_time,update_time) values
+(7,'3301081000000025',0,'172.16.171.219',2,0,0,'2018-08-20 13:46:42','2023-04-02 22:10:30');
+
+select last_car_no from t1 where id=1;
+select count(*) from t1;
+
+--source include/sync_slave_sql_with_master.inc
+--echo # on slave:
+sleep 1;
+select last_car_no from t1 where id=1;
+select count(*) from t1;
+
+connection master;
+--echo # on master:
+update t1 set last_car_no='JJY3786' where id=1;
+update t1 set last_car_no=NULL where id=1;
+update t1 set last_car_no='JJY3786' where id=1;
+update t1 set last_car_no=NULL where id=1;
+update t1 set last_car_no='JJY3786' where id=1;
+insert into t1(id,dev_code,port,server_ip,work_status,port_link_car,last_charge_mode,create_time,update_time) values
+(8,'3301081000000026',0,'172.16.171.219',2,0,0,'2018-08-20 13:46:42','2023-04-02 22:10:30');
+
+select last_car_no from t1 where id=1;
+select count(*) from t1;
+
+--source include/sync_slave_sql_with_master.inc
+--echo # on slave:
+sleep 1;
+select last_car_no from t1 where id=1;
+select count(*) from t1;
+
+connection master;
+--echo # on master:
+insert into t1(id,dev_code,port,server_ip,work_status,port_link_car,last_charge_mode,create_time,update_time) values
+(9,'3301081000000027',0,'172.16.171.219',2,0,0,'2018-08-20 13:46:42','2023-04-02 22:10:30');
+delete from t1 where id=9;
+insert into t1(id,dev_code,port,server_ip,work_status,port_link_car,last_charge_mode,create_time,update_time) values
+(9,'3301081000000027',0,'172.16.171.219',2,0,0,'2018-08-20 13:46:42','2023-04-02 22:10:30');
+select last_car_no from t1 where id=9;
+select count(*) from t1;
+
+--source include/sync_slave_sql_with_master.inc
+--echo # on slave:
+sleep 1;
+select last_car_no from t1 where id=9;
+select count(*) from t1;
+
+connection master;
+--echo # on master:
+update t1 set last_car_no=NULL where id=1;
+delete from t1 where id =1;
+select last_car_no from t1 where id=1;
+select count(*) from t1;
+
+--source include/sync_slave_sql_with_master.inc
+--echo # on slave:
+sleep 1;
+select last_car_no from t1 where id=1;
+select count(*) from t1;
+
+connection master;
+--echo # on master:
+insert into t1(id,dev_code,port,server_ip,work_status,port_link_car,last_charge_mode,create_time,update_time) values
+(10,'3301081000000028',0,'172.16.171.219',2,0,0,'2018-08-20 13:46:42','2023-04-02 22:10:30');
+update t1 set last_car_no='JJY3786' where id=10;
+select last_car_no from t1 where id=10;
+select count(*) from t1;
+
+--source include/sync_slave_sql_with_master.inc
+--echo # on slave:
+sleep 1;
+select last_car_no from t1 where id=10;
+select count(*) from t1;
+
+connection master;
+--echo # on master:
+insert into t1(id,dev_code,port,server_ip,work_status,port_link_car,last_charge_mode,create_time,update_time) values
+(11,'3301081000000029',0,'172.16.171.219',2,0,0,'2018-08-20 13:46:42','2023-04-02 22:10:30');
+update t1 set last_car_no='JJY3786' where id=11;
+delete from t1 where id =11;
+select last_car_no from t1 where id=11;
+select count(*) from t1;
+
+--source include/sync_slave_sql_with_master.inc
+--echo # on slave:
+sleep 1;
+select last_car_no from t1 where id=11;
+select count(*) from t1;
+
+connection master;
+--echo # on master:
+DROP DATABASE issue1523_test;
+--source include/sync_slave_sql_with_master.inc
+stop slave;

--- a/storage/tianmu/core/pack_int.cpp
+++ b/storage/tianmu/core/pack_int.cpp
@@ -32,7 +32,6 @@ namespace Tianmu {
 namespace core {
 PackInt::PackInt(DPN *dpn, PackCoordinate pc, ColumnShare *s) : Pack(dpn, pc, s) {
   is_real_ = ATI::IsRealType(s->ColType().GetTypeName());
-
   if (dpn_->NotTrivial()) {
     system::TianmuFile f;
     f.OpenReadOnly(s->DataFile());
@@ -177,7 +176,9 @@ void PackInt::ExpandOrShrink(uint64_t maxv, int64_t delta) {
 
 void PackInt::UpdateValueFloat(size_t locationInPack, const Value &v) {
   if (IsNull(locationInPack)) {
-    ASSERT(v.HasValue());
+    if (!v.HasValue())
+      return;
+    // ASSERT(v.HasValue(), col_share_->DataFile() + " locationInPack: " + std::to_string(locationInPack));
 
     // update null to non-null
     dpn_->synced = false;
@@ -344,9 +345,10 @@ void PackInt::DeleteByRow(size_t locationInPack) {
 
 void PackInt::UpdateValueFixed(size_t locationInPack, const Value &v) {
   if (IsNull(locationInPack)) {
-    ASSERT(v.HasValue());
-
-    // update null to non-null
+    if (!v.HasValue())
+      return;
+    // ASSERT(v.HasValue(), col_share_->DataFile() + " locationInPack: " + std::to_string(locationInPack));
+    //  update null to non-null
     dpn_->synced = false;
     UnsetNull(locationInPack);
 

--- a/storage/tianmu/core/pack_str.cpp
+++ b/storage/tianmu/core/pack_str.cpp
@@ -188,6 +188,11 @@ void PackStr::UpdateValue(size_t locationInPack, const Value &v) {
   dpn_->synced = false;
 
   if (IsNull(locationInPack)) {
+    // In the delta layer, there may be situations where null
+    // values are updated with null values, so when encountering this situation, it is directly returned
+    if (!v.HasValue())
+      return;
+
     // update null to non-null
 
     // first non-null value?
@@ -195,7 +200,7 @@ void PackStr::UpdateValue(size_t locationInPack, const Value &v) {
       dpn_->max_i = -1;
     }
 
-    ASSERT(v.HasValue());
+    // ASSERT(v.HasValue(), col_share_->DataFile() + " locationInPack: " + std::to_string(locationInPack));
     UnsetNull(locationInPack);
     dpn_->numOfNulls--;
     auto &str = v.GetString();

--- a/storage/tianmu/core/tianmu_attr.cpp
+++ b/storage/tianmu/core/tianmu_attr.cpp
@@ -1046,10 +1046,6 @@ void TianmuAttr::UpdateData(uint64_t row, Value &old_v, Value &new_v) {
 
   auto &dpn = get_dpn(pn);
   auto dpn_save = dpn;
-  if (dpn.Trivial()) {
-    // need to create pack struct for previous trivial pack
-    ha_tianmu_engine_->cache.GetOrFetchObject<Pack>(get_pc(pn), this);
-  }
 
   if (ct.Lookup() && new_v.HasValue()) {
     auto &str = new_v.GetString();
@@ -1079,11 +1075,12 @@ void TianmuAttr::UpdateData(uint64_t row, Value &old_v, Value &new_v) {
   ResetMaxMin(dpn);
 }
 
-void TianmuAttr::UpdateBatchData(core::Transaction *tx, const std::unordered_map<uint64_t, Value> &rows) {
+void TianmuAttr::UpdateBatchData(core::Transaction *tx,
+                                 const std::unordered_map<uint64_t, std::shared_ptr<Value>> &rows) {
   no_change = false;
 
   // group by pn
-  std::unordered_map<common::PACK_INDEX, std::unordered_map<uint64_t, Value>> packs;
+  std::unordered_map<common::PACK_INDEX, std::unordered_map<uint64_t, std::shared_ptr<Value>>> packs;
   for (const auto &row : rows) {
     auto row_id = row.first;
     auto row_val = row.second;
@@ -1092,7 +1089,7 @@ void TianmuAttr::UpdateBatchData(core::Transaction *tx, const std::unordered_map
     if (pack != packs.end()) {
       pack->second.emplace(row_id, row_val);
     } else {
-      packs.emplace(pn, std::unordered_map<uint64_t, Value>{{row_id, row_val}});
+      packs.emplace(pn, std::unordered_map<uint64_t, std::shared_ptr<Value>>{{row_id, row_val}});
     }
   }
 
@@ -1102,15 +1099,12 @@ void TianmuAttr::UpdateBatchData(core::Transaction *tx, const std::unordered_map
     CopyPackForWrite(pn);
     auto &dpn = get_dpn(pn);
     auto dpn_save = dpn;
-    if (dpn.Trivial()) {
-      ha_tianmu_engine_->cache.GetOrFetchObject<Pack>(get_pc(pn), this);
-    }
 
     for (const auto &row : pack.second) {
       uint64_t row_id = row.first;
-      Value row_val = row.second;
-      if (ct.Lookup() && row_val.HasValue()) {
-        auto &str = row_val.GetString();
+      std::shared_ptr<Value> row_val = row.second;
+      if (ct.Lookup() && row_val->HasValue()) {
+        auto &str = row_val->GetString();
         int code = m_dict->GetEncodedValue(str.data(), str.size());
         if (code < 0) {
           ASSERT(m_tx != nullptr, "attempt to update dictionary in readonly transaction");
@@ -1123,9 +1117,9 @@ void TianmuAttr::UpdateBatchData(core::Transaction *tx, const std::unordered_map
           }
           code = m_dict->Add(str.data(), str.size());
         }
-        row_val.SetInt(code);
+        row_val->SetInt(code);
       }
-      get_pack(pn)->UpdateValue(row2offset(row_id), row_val);
+      get_pack(pn)->UpdateValue(row2offset(row_id), *row_val);
     }
 
     dpn.synced = false;
@@ -1168,10 +1162,6 @@ void TianmuAttr::DeleteData(uint64_t row) {
   CopyPackForWrite(pn);
   auto &dpn = get_dpn(pn);
   auto dpn_save = dpn;
-  if (dpn.Trivial()) {
-    // need to create pack struct for previous trivial pack
-    ha_tianmu_engine_->cache.GetOrFetchObject<Pack>(get_pc(pn), this);
-  }
   get_pack(pn)->DeleteByRow(row2offset(row));
 
   // update global data
@@ -1201,9 +1191,6 @@ void TianmuAttr::DeleteBatchData(core::Transaction *tx, const std::vector<uint64
     CopyPackForWrite(pn);
     auto &dpn = get_dpn(pn);
     auto dpn_save = dpn;
-    if (dpn.Trivial()) {
-      ha_tianmu_engine_->cache.GetOrFetchObject<Pack>(get_pc(pn), this);
-    }
 
     for (const auto &row_id : pack.second) {
       get_pack(pn)->DeleteByRow(row2offset(row_id));
@@ -1260,9 +1247,6 @@ void TianmuAttr::CopyPackForWrite(common::PACK_INDEX pi) {
   // update current view
   m_idx[pi] = pos;
   auto &dpn(get_dpn(pi));
-
-  //  if (dpn.Trivial())
-  //      return;
 
   const PackCoordinate pc_old(m_tid, m_cid, m_share->GetPackIndex(&old_dpn));
   const PackCoordinate pc_new(get_pc(pi));

--- a/storage/tianmu/core/tianmu_attr.h
+++ b/storage/tianmu/core/tianmu_attr.h
@@ -92,7 +92,7 @@ class TianmuAttr final : public mm::TraceableObject, public PhysicalColumn, publ
 
   mm::TO_TYPE TraceableType() const override { return mm::TO_TYPE::TO_TEMPORARY; }
   void UpdateData(uint64_t row, Value &old_v, Value &new_v);
-  void UpdateBatchData(core::Transaction *tx, const std::unordered_map<uint64_t, Value> &rows);
+  void UpdateBatchData(core::Transaction *tx, const std::unordered_map<uint64_t, std::shared_ptr<Value>> &rows);
   void UpdateIfIndex(core::Transaction *tx, uint64_t row, uint64_t col, const Value &old_v, const Value &new_v);
   void Truncate();
   void DeleteData(uint64_t row);

--- a/storage/tianmu/core/tianmu_table.cpp
+++ b/storage/tianmu/core/tianmu_table.cpp
@@ -209,30 +209,10 @@ class DelayedInsertParser final {
             ptr += str_len;
           } break;
           case common::PackType::INT: {
-            if (attr->Type().Lookup()) {
-              uint32_t len = *(uint32_t *)ptr;
-              ptr += sizeof(uint32_t);
-              types::BString s(len == 0 ? "" : ptr, len);
-              int64_t *buf = reinterpret_cast<int64_t *>(vc.Prepare(sizeof(int64_t)));
-              *buf = attr->EncodeValue_T(s, true);
-              vc.ExpectedSize(sizeof(int64_t));
-              ptr += len;
-            } else {
-              int64_t *buf = reinterpret_cast<int64_t *>(vc.Prepare(sizeof(int64_t)));
-              *buf = *(int64_t *)ptr;
-
-              //              if (attr->GetIfAutoInc()) {
-              //                if (*buf == 0)  // Value of auto inc column was not assigned by user
-              //                  *buf = attr->AutoIncNext();
-              //                if (static_cast<uint64_t>(*buf) > attr->GetAutoInc()) {
-              //                  if (*buf > 0 || ((attr->TypeName() == common::ColumnType::BIGINT) &&
-              //                  attr->GetIfUnsigned()))
-              //                    attr->SetAutoInc(*buf);
-              //                }
-              //              }
-              vc.ExpectedSize(sizeof(int64_t));
-              ptr += sizeof(int64_t);
-            }
+            int64_t *buf = reinterpret_cast<int64_t *>(vc.Prepare(sizeof(int64_t)));
+            *buf = *(int64_t *)ptr;
+            vc.ExpectedSize(sizeof(int64_t));
+            ptr += sizeof(int64_t);
           } break;
           default:
             break;
@@ -280,10 +260,10 @@ class DelayedUpdateParser final {
                       std::shared_ptr<index::TianmuTableIndex> index)
       : attrs(attrs), update_rows(update_rows), index_table(std::move(index)) {}
 
-  uint GetRows(std::vector<std::unordered_map<uint64_t, Value>> &update_cols_buf) {
+  uint GetRows(std::vector<std::unordered_map<uint64_t, std::shared_ptr<Value>>> &update_cols_buf) {
     update_cols_buf.reserve(attrs.size());
     for (int i = 0; i < attrs.size(); i++) {
-      update_cols_buf.emplace_back(std::unordered_map<uint64_t, Value>());
+      update_cols_buf.emplace_back(std::unordered_map<uint64_t, std::shared_ptr<Value>>());
     }
     uint update_row_num = update_rows->size();
     for (auto &[row_id, row] : *update_rows) {
@@ -291,7 +271,7 @@ class DelayedUpdateParser final {
       DeltaRecordHeadForUpdate rec_head;
       row_ptr = rec_head.recordDecode(row_ptr);
       for (uint col_id = 0; col_id < attrs.size(); col_id++) {
-        core::Value val;
+        std::shared_ptr<Value> val = std::make_shared<Value>();
         if (!rec_head.update_mask_[col_id]) {
           continue;
         }
@@ -304,29 +284,21 @@ class DelayedUpdateParser final {
         switch (attr->GetPackType()) {
           case common::PackType::STR: {
             uint32_t str_len = rec_head.field_len_[col_id];
-            val.SetString(const_cast<char *>(row_ptr), str_len);
+            val->SetString(const_cast<char *>(row_ptr), str_len);
             update_cols_buf[col_id].emplace(row_id, val);
             row_ptr += str_len;
           } break;
           case common::PackType::INT: {
-            if (attr->Type().Lookup()) {
-              uint32_t len = *(uint32_t *)row_ptr;
-              row_ptr += sizeof(uint32_t);
-              types::BString s(len == 0 ? "" : row_ptr, len);
-              int64_t int_val = attr->EncodeValue_T(s, true);
-              row_ptr += len;
-            } else {
-              int64_t int_val = *(int64_t *)row_ptr;
-              if (attr->GetIfAutoInc()) {
-                if (int_val == 0)  // Value of auto inc column was not assigned by user
-                  int_val = attr->AutoIncNext();
-                if (static_cast<uint64_t>(int_val) > attr->GetAutoInc())
-                  attr->SetAutoInc(int_val);
-              }
-              val.SetInt(int_val);
-              update_cols_buf[col_id].emplace(row_id, val);
-              row_ptr += sizeof(int64_t);
+            int64_t int_val = *(int64_t *)row_ptr;
+            if (attr->GetIfAutoInc()) {
+              if (int_val == 0)  // Value of auto inc column was not assigned by user
+                int_val = attr->AutoIncNext();
+              if (static_cast<uint64_t>(int_val) > attr->GetAutoInc())
+                attr->SetAutoInc(int_val);
             }
+            val->SetInt(int_val);
+            update_cols_buf[col_id].emplace(row_id, val);
+            row_ptr += sizeof(int64_t);
           } break;
           default:
             break;
@@ -1659,7 +1631,8 @@ int TianmuTable::AsyncParseUpdateRecords(system::IOParameters *iop,
   clock_gettime(CLOCK_REALTIME, &t1);
   auto index_table = ha_tianmu_engine_->GetTableIndex(share->Path());
   DelayedUpdateParser parser(m_attrs, update_records, index_table);
-  std::vector<std::unordered_map<uint64_t, Value>> update_cols_buf;
+  std::vector<std::unordered_map<uint64_t, std::shared_ptr<Value>>> update_cols_buf;
+
   int returned_row_num = parser.GetRows(update_cols_buf);
   if (returned_row_num > 0) {
     utils::result_set<void> res;

